### PR TITLE
Support date_sub function

### PR DESCRIPTION
**Pull request by zimulala**
#419

### CODE REVIEW

**Comment by siddontang** [comment]
What does Cast mean?

**Comment by zimulala** [comment]
here means calculation

**Comment by siddontang** [comment]
cast is not for calculation, it means conversation.

**Comment by zimulala** [comment]
It has calculated the meanings in the dictionary, here it may not be an appropriate name. Do you have a suggestion?

**Comment by siddontang** [comment]
DateArith or DateArithmetic

**Comment by zimulala** [comment]
gotcha, `DateCast` will be renamed `DateArith`.

File: parser/parser.y [hunk 1]
```diff
@@ -2285,12 +2286,22 @@ FunctionCallNonKeyword:
 	}
 |	"DATE_ADD" '(' Expression ',' "INTERVAL" Expression TimeUnit ')'
 	{
-		$$ = &expression.DateAdd{
+		$$ = &expression.DateArith{
+			Op:"ADD",
```
Comment (siddontang) [inline parser/parser.y]: use a const int value instead of string, like

const (
  DateAdd 
  DateSub
)

Comment (zimulala) [inline parser/parser.y]: got it.

**Comment by zimulala** [comment]
PTAL

File: expression/date_arith.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/util/types"
+)
+
+const (
+	// DateAdd is to run date_add function option.
+	// See: https://dev.mysql.com/doc/refman/5.7/en/date-and-time-functions.html#function_date-add
+	DateAdd = 1
```
Comment (qiuyesuifeng) [inline expression/date_arith.go]: I prefer to use iota, like 
```
type xxxType byte

const (
       DateAdd  xxxType = iota +1
       DateSub
)
```

File: expression/date_arith.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+// Accept implements the Visitor Accept interface.
+func (da *DateArith) Accept(v Visitor) (Expression, error) {
+	return v.VisitDateArith(da)
+}
+
+// String implements the Expression String interface.
+func (da *DateArith) String() string {
+	var str string
+	if da.isAdd() {
```
Comment (qiuyesuifeng) [inline expression/date_arith.go]: Seems no need to extract a function, it is only one line.

File: expression/date_arith.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	return t, nil
+}
+
+func (da *DateArith) evalArgs(ctx context.Context, args map[interface{}]interface{}) (
+	mysql.Time, int64, int64, int64, time.Duration, error) {
+	dVal, err := da.Date.Eval(ctx, args)
+	if dVal == nil || err != nil {
+		return mysql.ZeroTimestamp, 0, 0, 0, 0, errors.Trace(err)
+	}
+	dValStr, err := types.ToString(dVal)
```
Comment (qiuyesuifeng) [inline expression/date_arith.go]: I suggest `Adding some necessary empty space for readability and for eye-health...` which refer from xiang90 in https://github.com/reborndb/reborn/pull/50

File: expression/date_arith_test.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+	. "github.com/pingcap/check"
+	"github.com/pingcap/tidb/mysql"
+)
+
+var _ = Suite(&testDateArithSuite{})
+
+type testDateArithSuite struct {
+}
+
+func (t *testDateArithSuite) TestDateArith(c *C) {
```
Comment (qiuyesuifeng) [inline expression/date_arith_test.go]: Please some test for `DateSub`.

File: expression/date_arith.go (hunk 2 above)
Comment (siddontang) [inline expression/date_arith.go]: +1
Btw, you can  implement String() for DateArithType.

Comment (zimulala) [inline expression/date_arith.go]: gotcha

File: expression/date_arith.go (hunk 3 above)
Comment (zimulala) [inline expression/date_arith.go]: I think it's clear.

File: expression/date_arith_test.go (hunk 5 above)
Comment (zimulala) [inline expression/date_arith_test.go]: Here are.

Comment (qiuyesuifeng) [inline expression/date_arith_test.go]: Yes, gotcha.

File: expression/date_arith.go (hunk 4 above)
Comment (zimulala) [inline expression/date_arith.go]: Your're right, but I don't think it's necessary so many blank lines.

File: expression/visitor.go [hunk 6]
```diff
@@ -107,8 +107,8 @@ type Visitor interface {
 	// VisitFunctionTrim visits FunctionTrim expression.
 	VisitFunctionTrim(v *FunctionTrim) (Expression, error)
 
-	// VisitDateAdd visits DateAdd expression.
-	VisitDateAdd(da *DateAdd) (Expression, error)
+	// VisitDateArith visits DateAdd and DateSub expression.
```
Comment (qiuyesuifeng) [inline expression/visitor.go]: This comment can update to `VisitDateArith visits DateArith expression.`, for we only have DateArith expression now.

Comment (zimulala) [inline expression/visitor.go]: got it.

File: expression/date_arith.go (hunk 4 above)
Comment (siddontang) [inline expression/date_arith.go]: a black line here looks better to distinguish above logic unit.  :-)

Comment (zimulala) [inline expression/date_arith.go]: But it used here, other place also want to use, then too much.
And why want to use it here?

Comment (qiuyesuifeng) [inline expression/date_arith.go]: It is not a big problem, only a suggestion, u can do what u like :)

Comment (zimulala) [inline expression/date_arith.go]: thanks :)

**Comment by zimulala** [comment]
PTAL @siddontang

**Comment by siddontang** [comment]
LGTM

**Pull request merged**